File: 1. Two Sum/two_sum.ts
function twoSum(nums: number[], target: number): number[] {
  const indeces: number[] = [];

  for (let i: number = 0; i <= nums.length; i++) {
    for (let j: number = i; j <= nums.length; j++) {
      if (i === j) continue;

      if (target === nums[i] + nums[j]) {
        indeces.push(i, j);
        return indeces;
      }  
    }
  }

  return indeces;
}



function twoSum2(nums: number[], target: number): number[] {
  const indeces: number[] = [];
  const map = new Map<number, number>();

  nums.forEach((n, i) => map.set(n, i));

  for (let i: number = 0; i <= nums.length; i++) {
    let difference: number = target - nums[i];
    if (map.has(difference)) {
      const j: number | undefined = map.get(difference);

      if (j && i !== j) {
        indeces.push(i, j);
        break;
      }
    }
  }

  return indeces;
}

function twoSum3(nums: number[], target: number): number[] {
  const indeces: number[] = [];
  const map = new Map<number, number>();

  for (let i: number = 0; i <= nums.length; i++) {
    let difference: number = target - nums[i];
    if (map.has(difference)) {
      const j: number | undefined = map.get(difference);

      if (typeof j === "number") {
        indeces.push(i, j);
        break;
      }
    } else {
      map.set(nums[i], i);
    }
  }

  return indeces;
}

File: 1. Two Sum/two_sum.rb
def two_sum(nums, target)
  indeces = []
  step = 0

  while indeces.empty? && step != nums.size
    nums[(step + 1)..-1].each_with_index do |num, i|
 
      if target == (nums[step] + num)
       indeces << step << (i + step + 1)
      end
    end

    step += 1
  end

  indeces
end


def two_sum(nums, target)
  nums.each_with_index do |x, i|
    nums.each_with_index do |y, j|
      next if i == j

      return [i, j] if target == x + y
    end
  end
end

def two_sum(nums, target)
  hash = Hash.new
  
  nums.each_with_index &-> n, i { hash[n] = i }

  nums.each_with_index do |n, i|
    difference = target - n

    return [i, hash[difference]] if hash.include?(difference) && i != hash[difference]
  end
end

def two_sum(nums, target)
  hash = Hash.new

  nums.each_with_index do |n, i|
    difference = target - n

    return [i, hash[difference]] if hash.include? difference

    hash[n] = i
  end
end
